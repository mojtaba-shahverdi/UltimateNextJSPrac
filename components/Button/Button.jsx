import React from 'react'
import styled from '@emotion/styled'
import { useTheme } from '@emotion/react'

// eslint-disable-next-line react/prop-types
const Button = ({ type = 'button', children, widthAll, loading }) => {
  const theme = useTheme()

  const Layout = styled.div`
    width: 100%;
  `
  const ButtonEl = styled.button`
    width: ${widthAll ? '100%' : 'auto'};
    background: ${theme.colors.primary};
    color: #fff;
    font-size: ${theme.typography.paragraph.fontSize};
    text-decoration: none;
    padding: ${theme.spaces[4]};
    border: ${theme.border[1]};
    border-radius: ${theme.borderRadius[1]};
    display: inline-block;
    position: relative;
    cursor: pointer;
    transition: all 0.4s ease-in-out;
    &:hover {
      box-shadow: ${theme.boxShadow.light};
      background: ${theme.colors.primaryHover};
      color: #000;
      font-weight: bold;
    }
    &:active {
      box-shadow: ${theme.boxShadow.light} inset;
    }
  `
  return (
    <Layout>
      <ButtonEl>
        {loading && <img src="/loading.gif" width="6%" />}
        {!loading && children}
      </ButtonEl>
    </Layout>
  )
}

export default Button
