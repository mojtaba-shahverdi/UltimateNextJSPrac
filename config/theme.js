const theme = {
  colors: {
    primary: '#00cec9',
    primaryHover: '#00fffb',
    backgroundPrimary: '#f8f9fa',
  },
  typography: {
    paragraph: {
      fontSize: 1.266,
      lineHeight: 1.15,
      fontWeight: 300,
    },
    h1: {
      fontSize: 1.802,
      lineHeight: 1.15,
      fontWeight: 500,
    },
    h2: {
      fontSize: 1.602,
      lineHeight: 1.15,
      fontWeight: 500,
    },
    h3: {
      fontSize: 1.424,
      lineHeight: 1.15,
      fontWeight: 500,
    },
    h4: {
      fontSize: 1.125,
      lineHeight: 1.15,
      fontWeight: 500,
    },
    small: {
      fontSize: 0.889,
      lineHeight: 1.15,
      fontWeight: 100,
    },
  },
  spaces: ['0px', '2px', '4px', '8px', '16px', '32px'],
  borderRadius: ['0px', '2px'],
  border: ['0px', '2px'],
  borderColor: {
    light: '#e3e4e6',
  },
  boxShadow: {
    light: '0 4px 16px rgba(0, 1, 31, 0.2)',
  },
}

export default theme
