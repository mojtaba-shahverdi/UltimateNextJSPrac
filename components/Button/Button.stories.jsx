import React from 'react'
import Button from './Button'

export default {
  title: 'Button',
  component: Button,
}

export const simple = () => <Button>Click here</Button>
export const simpleWidthAll = () => <Button widthAll>Click here</Button>
