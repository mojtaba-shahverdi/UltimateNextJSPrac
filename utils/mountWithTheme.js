import React from 'react'
import { ThemeProvider } from '@emotion/react'
import theme from '../config/theme'
import renderer from 'react-test-renderer'

const mountWithTheme = (children) => {
    return renderer.create(
        <ThemeProvider theme={theme}>{children}</ThemeProvider>
    )
}

export default mountWithTheme
