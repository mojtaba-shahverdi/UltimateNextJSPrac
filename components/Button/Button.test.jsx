/* eslint-disable prettier/prettier */
import React from 'react'
import { fireEvent, render } from '@testing-library/react';
import Button from './Button'
import mountWithTheme from '../../utils/mountWithTheme'
import { ThemeProvider } from '@emotion/react';
import theme from '../../config/theme';

describe('Button', () => {
    test('Render Correctly', () => {
        const component = mountWithTheme(<Button>Sample Button</Button>)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
    test('Shows Loading State', () => {
        const component = mountWithTheme(<Button loading>Sample Button with Loading</Button>)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
    test('Shows Disabled State', () => {
        const component = mountWithTheme(<Button Disabled>Sample Button with Disabled State</Button>)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
    test('Shows buttonType Secondary', () => {
        const component = mountWithTheme(<Button buttonType="secondary">Sample Button with Secondary Style</Button>)
        const tree = component.toJSON()
        expect(tree).toMatchSnapshot()
    })
    test('Renders in DOM correctly', () => {
        const { getByText } = render(
            <ThemeProvider theme={theme}>
                <Button>Sample Button in DOM</Button>
            </ThemeProvider>
        )
        expect(getByText(/Sample Button/)).toBeTruthy()
    })
    test('onClick works Fine', () => {
        let variable = 1
        const { getByText } = render(
            <ThemeProvider theme={theme}>
                <Button onClick={() => {
                    variable++
                }}>Sample Button in DOM</Button>
            </ThemeProvider>
        )
        fireEvent.click(getByText(/Sample Button/))

        expect(variable).toBe(2)
    })
})
