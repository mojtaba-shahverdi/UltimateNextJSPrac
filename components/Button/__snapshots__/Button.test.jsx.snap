// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button Render Correctly 1`] = `
[
  <style
    dangerouslySetInnerHTML={
      {
        "__html": ".css-ybrhvy{width:100%;}",
      }
    }
    data-emotion="css ybrhvy"
  />,
  <div
    className="css-ybrhvy"
  >
    <style
      dangerouslySetInnerHTML={
        {
          "__html": ".css-1bn5aq7{width:auto;background:#00cec9;color:#fff;font-size:1;-webkit-text-decoration:none;text-decoration:none;padding:16px;border:2px;border-radius:2px;display:inline-block;position:relative;cursor:pointer;-webkit-transition:all 0.4s ease-in-out;transition:all 0.4s ease-in-out;}.css-1bn5aq7:hover{box-shadow:0 4px 16px rgba(0, 1, 31, 0.2);background:#00fffb;color:#000;font-weight:bold;}.css-1bn5aq7:active{box-shadow:0 4px 16px rgba(0, 1, 31, 0.2) inset;}",
        }
      }
      data-emotion="css 1bn5aq7"
    />
    <button
      className="css-1bn5aq7"
    >
      Sample Button
    </button>
  </div>,
]
`;
